### PR TITLE
feat: integrate a secret manager

### PR DESCRIPTION
**Pull request by Chethan-rao**
This PR integrates secret manager and makes use of it to store the entity dek.
Keeping in mind about the old dek's, this PR for now calls secret manager to store dek, only when new merchant hits add request. This is done to keep the number of merchant dek's to be migrated in future fixed. 
During add request, if new merchant, create a merchant dek in merchant table and also call secret manager to transfer the created dek to the service.

### CODE REVIEW

File: config.example.toml [hunk 1]
```diff
@@ -59,6 +59,17 @@ hyperswitch = { master_key = "feffe9928665731c6d6a8f9467308308feffe9928665731c6d
...
+# Api client
+[api_client]
+client_idle_timeout = 90 # timeout for idle sockets being kept-alive
+pool_max_idle_per_host = 10 # maximum idle connection per host allowed in the pool.
+identity = "" # identity to be used for client certificate authentication in mtls.
+
+# Configuration for the Key Manager Service
+[key_manager]
+url = "http://localhost:5000" # URL of the encryption service
+cert = "" # Represents a server X509 certificate for mtls
```
Comment (NishantJoshi00) [inline config.example.toml:75]: Should we take the certificate as the input? or a file path instead?

File: src/api_client.rs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            .redirect(reqwest::redirect::Policy::none())
+            .pool_idle_timeout(std::time::Duration::from_secs(
+                global_config.api_client.client_idle_timeout,
+            ))
+            .pool_max_idle_per_host(global_config.api_client.pool_max_idle_per_host);
+
+        #[cfg(feature = "keymanager_mtls")]
+        {
+            let client_identity = reqwest::Identity::from_pem(
+                global_config.api_client.identity.clone().peek().as_ref(),
```
Comment (NishantJoshi00) [inline src/api_client.rs:78]: Is this `.clone()` required? if we are following it with `.peek().as_ref()`

File: src/api_client.rs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            )
+            .change_error(error::ApiClientError::IdentityParseFailed)?;
+
+            let key_manager_cert = reqwest::Certificate::from_pem(
+                global_config.key_manager.cert.clone().peek().as_ref(),
```
Comment (NishantJoshi00) [inline src/api_client.rs:83]: same here

File: src/crypto/keymanager.rs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                    .db
+                    .insert_entity(entity_id, &response.identifier.get_identifier())
+                    .await?)
+            }
+            false => Err(KeyManagerError::DbError.into()),
+        },
+    }
+}
+
+pub async fn transfer_key_to_key_manager(
```
Comment (NishantJoshi00) [inline src/crypto/keymanager.rs:57]: can you add comments, (as this function will be removed in the future, right?)

File: src/routes/key_custodian.rs [hunk 5]
```diff
@@ -114,12 +114,15 @@ pub async fn decrypt(
...
             aes_decrypt_custodian_key(&mut tenant_config, inner_key1, inner_key2).await?;
 
-            let tenant_app_state =
-                TenantAppState::new(global_app_state.global_config.clone(), tenant_config)
-                    .await
-                    .change_context(error::ApiError::TenantError(
-                        "Failed while creating AppState for tenant",
-                    ))?;
+            let tenant_app_state = TenantAppState::new(
+                global_app_state.global_config.clone(),
```
Comment (NishantJoshi00) [inline src/routes/key_custodian.rs:118]: The global config is starting to become bulky, we should wrap it with a shared reference.

**Review by NishantJoshi00 — CHANGES_REQUESTED** [review]
LGTM

File: config.example.toml (hunk 1 above)
Comment (Chethan-rao) [reply config.example.toml:75]: which one is convenient and secure in deployment environments? if we pass direct certificate, then we'll have to kms encrypt it. But if its file path then the file has to be kept secure. what do u suggest?

File: src/crypto/keymanager.rs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    )
+    .await?;
+
+    Ok(response.data)
+}
+
+pub async fn decrypt_data_using_key_manager<T>(
+    state: &TenantAppState,
+    request_body: DataDecryptionRequest,
+) -> Result<T, ContainerError<KeyManagerError>>
```
Comment (NishantJoshi00) [inline src/crypto/keymanager.rs:100]: can we have it return `Secret<T>`? Its' supposedly sensitive data that is decrypted by that service, right?

File: src/crypto/keymanager.rs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    state: &TenantAppState,
+    url: String,
+    request_body: T,
+) -> Result<R, ContainerError<E>>
+where
+    T: serde::Serialize + Send + Sync + 'static,
+    R: serde::de::DeserializeOwned,
+    ContainerError<E>: From<ContainerError<ApiClientError>> + Send + Sync,
+{
+    let headers = [("Content-type".into(), "application/json".into())]
```
Comment (NishantJoshi00) [inline src/crypto/keymanager.rs:127]: can you use the `CONTENT_TYPE` const that you have created?

File: src/crypto/keymanager/types.rs [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+pub struct DataEncryptionRequest {
+    #[serde(flatten)]
+    pub identifier: Identifier,
+    pub data: DecryptedData,
+}
+
+impl DataEncryptionRequest {
+    pub fn create_request<T>(
+        key_identifier: String,
+        data: &T,
```
Comment (NishantJoshi00) [inline src/crypto/keymanager/types.rs:62]: can this be a `Secret` too?

**Review by NishantJoshi00 — CHANGES_REQUESTED** [review]

**Review by dracarys18 — APPROVED** [review]

**Review by ShankarSinghC — APPROVED** [review]

**Review by NishantJoshi00 — APPROVED** [review]

**Pull request merged**